## STAT 306 Finding Relationships in Data

### Group Project Report

Group A2 - Tran Anh Thu Phung, Ena Gupta, Yihang He, Haiyue Lu

### Import

In [1]:
library(tidyverse)
library(repr)
library(broom)
library(leaps)
library(moderndive)
library(gridExtra)
library(grid)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.3     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors

Attaching package: ‘gridExtra’


The following object is masked from ‘package:dplyr’:

    combine




### 1. Data:
**1.1 Introduction:**

Sleep is a fundamental aspect of human life, playing a critical role in our overall health and well-being. In today’s fast-paced world, many individuals struggle with sleep-related issues, such as insomnia, sleep disturbances, and reduced sleep efficiency. To understand how lifestyle factors impact sleep efficiency, in this project, we will use a data set from Kaggle to help individuals make informed choices to improve their sleep quality and overall health.

**1.2 Loading the Sleep Efficiency data**

To read the dataset in R, we will read the dataset from the web into R. 

In [2]:
url <- "https://raw.githubusercontent.com/Wendy1907/Sleep_Efficiency_Project/main/Sleep_Efficiency.csv"
df <- read_csv(url)
head(df)

Rows: 452 Columns: 15
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr   (2): Gender, Smoking status
dbl  (11): ID, Age, Sleep duration, Sleep efficiency, REM sleep percentage, ...
dttm  (2): Bedtime, Wakeup time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


ID Age Gender Bedtime             Wakeup time         Sleep duration
1 1  65  Female 2021-03-06 01:00:00 2021-03-06 07:00:00 6.0           
2 2  69  Male   2021-12-05 02:00:00 2021-12-05 09:00:00 7.0           
3 3  40  Female 2021-05-25 21:30:00 2021-05-25 05:30:00 8.0           
4 4  40  Female 2021-11-03 02:30:00 2021-11-03 08:30:00 6.0           
5 5  57  Male   2021-03-13 01:00:00 2021-03-13 09:00:00 8.0           
6 6  36  Female 2021-07-01 21:00:00 2021-07-01 04:30:00 7.5           
  Sleep efficiency REM sleep percentage Deep sleep percentage
1 0.88             18                   70                   
2 0.66             19                   28                   
3 0.89             20                   70                   
4 0.51             23                   25                   
5 0.76             27                   55                   
6 0.90             23                   60                   
  Light sleep percentage Awakenings Caffeine consumption Alcohol consumption
1 12                     0           0                   0                  
2 53                     3           0                   3                  
3 10                     1           0                   0                  
4 52                     3          50                   5                  
5 18                     3           0                   3                  
6 17                     0          NA                   0                  
  Smoking status Exercise frequency
1 Yes            3                 
2 Yes            3                 
3 No             3                 
4 Yes            1                 
5 No             3                 
6 No             1

**1.3 Data Summary**

The dataset contains **452 observations** in total with **15 columns** including ID.

| Variable Name      | Type of Variable | Explanation |
| ----------- | :-----------: | ----------- |
| `ID`     | int       | A unique identifier for each test subject. |
| `Age`   | int        | The age of the test subjects with a minimum value of 9 years old, a maximum value of 69 years old, and a mean of about 40 years old.|
| `Gender`   | chr       | The gender of the test subject, e.g., Male or Female. |
| `Bedtime`   | chr        |Indicate when each subject goes to bed, with the date and time.|
| `Wakeup time`   | chr        | Indicate when each subject wakes up each day, with the date and time.|
| `Sleep duration`   | dbl        | Records the total amount of time each subject slept in hours, with a minimum value of 5 hours, a maximum duration of 10 hours, and an average of 7.47 hours.|
| `Sleep efficiency`   | dbl       | A measure of the proportion of time spent in bed that is spent asleep, with a minimum weight of 50%, maximum of 99% and average of 78.89%.|
| `REM sleep percentage`   | int        | Amount of time (in percentage) each subject spent in the stage of REM sleep. Minimum of 15%, maximum of 30% and mean of 22.61%.|
| `Deep sleep percentage`   | int        | Amount of time (in percentage) each subject spent in the stage of deep sleep. Minimum of 18%, maximum of 75% and mean of 52.82%.|
| `Light sleep percentage`   | int        | Amount of time (in percentage) each subject spent in the stage of light sleep. Minimum of 7%, maximum of 63% and mean of 24.56%.|
| `Awakenings`   | dbl        | Record the number of times each subject wakes up during the night. There are a total of 20 null variables, with a minimum of 0 times, a maximum of 4 times, and a mean of 1.64.|
| `Caffeine consumption`   | dbl        | The amount of caffeine consumed in the 24 hours before bedtime (in mg), with 25 null variables. |
| `Alcohol consumption`   | dbl        | The amount of alcohol consumed in the 24 hours before bedtime (in oz), with 14 null variables. |
| `Smoking status`   | chr        | The smoking status of the test subject, whether they smoke or not. |
| `Exercise frequency`   | dbl        | The number of times the test subject exercises each week. With 6 null variables, with a mean of 1.79, minimum of 0 and maximum of 5 times a week. |

### 2. Research Question:

Can sleep efficiency be improved through lifestyle interventions(caffeine consumption, alcohol consumption, smoking status, exercise frequency), and what are the most effective strategies for enhancing sleep quality?

### 3. Cleaning and Warngling the dataset:

#### 3.1 Cleaning the dataset:

**3.1.1 Rename columns**

Since the columns name have space in the middle, then, we decided to rename them.

In [3]:
df <- df %>% 
        rename("Wakeup.time" = "Wakeup time",
               "Sleep.duration" = "Sleep duration",
               "Sleep.efficiency" = "Sleep efficiency",
               "REM.sleep.percentage" = "REM sleep percentage",
               "Deep.sleep.percentage" = "Deep sleep percentage",
               "Light.sleep.percentage" = "Light sleep percentage",
               "Caffeine.consumption" = "Caffeine consumption",
               "Alcohol.consumption" = "Alcohol consumption",
               "Smoking.status" = "Smoking status",
               "Exercise.frequency" = "Exercise frequency"
              )
head(df)

ID Age Gender Bedtime             Wakeup.time         Sleep.duration
1 1  65  Female 2021-03-06 01:00:00 2021-03-06 07:00:00 6.0           
2 2  69  Male   2021-12-05 02:00:00 2021-12-05 09:00:00 7.0           
3 3  40  Female 2021-05-25 21:30:00 2021-05-25 05:30:00 8.0           
4 4  40  Female 2021-11-03 02:30:00 2021-11-03 08:30:00 6.0           
5 5  57  Male   2021-03-13 01:00:00 2021-03-13 09:00:00 8.0           
6 6  36  Female 2021-07-01 21:00:00 2021-07-01 04:30:00 7.5           
  Sleep.efficiency REM.sleep.percentage Deep.sleep.percentage
1 0.88             18                   70                   
2 0.66             19                   28                   
3 0.89             20                   70                   
4 0.51             23                   25                   
5 0.76             27                   55                   
6 0.90             23                   60                   
  Light.sleep.percentage Awakenings Caffeine.consumption Alcohol.consumption
1 12                     0           0                   0                  
2 53                     3           0                   3                  
3 10                     1           0                   0                  
4 52                     3          50                   5                  
5 18                     3           0                   3                  
6 17                     0          NA                   0                  
  Smoking.status Exercise.frequency
1 Yes            3                 
2 Yes            3                 
3 No             3                 
4 Yes            1                 
5 No             3                 
6 No             1

**3.1.2 Finding Null variables**

First, we have to check if the data contains any null or `N/A` value or not

In [4]:
sum(is.na(df))

[1] 65

The function `is.na` which will identify the null value giving out the result that there are 65 null variables in our dataset. We want to take a look deeper to see which columns contain these null variables.

In [5]:
colSums(is.na(df))

ID                    Age                 Gender 
                     0                      0                      0 
               Bedtime            Wakeup.time         Sleep.duration 
                     0                      0                      0 
      Sleep.efficiency   REM.sleep.percentage  Deep.sleep.percentage 
                     0                      0                      0 
Light.sleep.percentage             Awakenings   Caffeine.consumption 
                     0                     20                     25 
   Alcohol.consumption         Smoking.status     Exercise.frequency 
                    14                      0                      6

Based on the summary above, we can see that there are total 20 null variables in `Awakenings` columns, 25 in `Caffeine consumption`, 14 in `Alcohol consumption`, and 6 in `Exercise frequency` in total. Then, we decided to remove those null values from our dataset.

In [6]:
df_clean <- df %>% drop_na(Awakenings, Caffeine.consumption, Alcohol.consumption, Exercise.frequency)

**3.1.3 Finding duplicate values**

Next step, we are going to check whether the data have filtered null values have duplicate value or not.

In [7]:
sum(duplicated(df_clean))

[1] 0

In this dataset, we can see that there is no duplicate value.

Then, we have got a clean dataset. The clean dataset that we have gained is contain total of **388 observations** and **15 different columns**.

In [8]:
head(df_clean)

ID Age Gender Bedtime             Wakeup.time         Sleep.duration
1 1  65  Female 2021-03-06 01:00:00 2021-03-06 07:00:00 6             
2 2  69  Male   2021-12-05 02:00:00 2021-12-05 09:00:00 7             
3 3  40  Female 2021-05-25 21:30:00 2021-05-25 05:30:00 8             
4 4  40  Female 2021-11-03 02:30:00 2021-11-03 08:30:00 6             
5 5  57  Male   2021-03-13 01:00:00 2021-03-13 09:00:00 8             
6 7  27  Female 2021-07-21 21:00:00 2021-07-21 03:00:00 6             
  Sleep.efficiency REM.sleep.percentage Deep.sleep.percentage
1 0.88             18                   70                   
2 0.66             19                   28                   
3 0.89             20                   70                   
4 0.51             23                   25                   
5 0.76             27                   55                   
6 0.54             28                   25                   
  Light.sleep.percentage Awakenings Caffeine.consumption Alcohol.consumption
1 12                     0           0                   0                  
2 53                     3           0                   3                  
3 10                     1           0                   0                  
4 52                     3          50                   5                  
5 18                     3           0                   3                  
6 47                     2          50                   0                  
  Smoking.status Exercise.frequency
1 Yes            3                 
2 Yes            3                 
3 No             3                 
4 Yes            1                 
5 No             3                 
6 Yes            1

#### 3.2 Wrangling the Sleep Efficiency data

Since our research question focus on factors such as `Caffeine.consumption`, `Alcohol.consumption`, `Smoking.status`, `Exercise.frequency`, and the target `Sleep.efficiency`, then we only interested in those columns to finding the most effective strategies enhancing sleep quality.

In [9]:
data <- df_clean %>%
select(Caffeine.consumption, Alcohol.consumption, Smoking.status, Exercise.frequency, Sleep.efficiency)
head(data)

Caffeine.consumption Alcohol.consumption Smoking.status Exercise.frequency
1  0                   0                   Yes            3                 
2  0                   3                   Yes            3                 
3  0                   0                   No             3                 
4 50                   5                   Yes            1                 
5  0                   3                   No             3                 
6 50                   0                   Yes            1                 
  Sleep.efficiency
1 0.88            
2 0.66            
3 0.89            
4 0.51            
5 0.76            
6 0.54